File: mobile/screens/HomeScreen.js
/**
 * Nom Auteur : Dr Abdelhakim Kellouche
 * Nom de l'application : Plateforme de Pilotage du Réseau d'Assainissement
 * Numéro version : 1.0.0
 * Date de création : 30-05-2026
 * Date de modification : 30-05-2026
 * 
 * Objectif du module :
 * Écran d'accueil principal de l'application mobile. Fournit l'état de connexion, 
 * un interrupteur de mode hors ligne, un badge des modifications locales en attente, 
 * et la navigation vers les modules cartographiques et de synchronisation.
 */

import React, { useEffect, useState } from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  ScrollView,
  Alert,
  Switch,
} from 'react-native';
import { healthService } from '../services/api';
import { storageService } from '../services/storage';

export default function HomeScreen({ navigation }) {
  const [networkMode, setNetworkMode] = useState('online');
  const [serverStatus, setServerStatus] = useState('checking');
  const [serverInfo, setServerInfo] = useState(null);
  const [pendingCount, setPendingCount] = useState(0);

  useEffect(() => {
    loadSettings();
    const unsubscribe = navigation.addListener('focus', () => {
      loadSettings();
    });
    return unsubscribe;
  }, [navigation]);

  const loadSettings = async () => {
    try {
      const mode = await storageService.getNetworkMode();
      setNetworkMode(mode);
      
      const changes = await storageService.getPendingChanges();
      setPendingCount(changes.length);

      if (mode === 'online') {
        checkServerStatus();
      } else {
        setServerStatus('offline');
      }
    } catch (e) {
      console.error(e);
    }
  };

  const toggleNetworkMode = async (value) => {
    const newMode = value ? 'online' : 'offline';
    setNetworkMode(newMode);
    await storageService.setNetworkMode(newMode);
    if (newMode === 'online') {
      checkServerStatus();
    } else {
      setServerStatus('offline');
      setServerInfo(null);
    }
  };

  const checkServerStatus = async () => {
    try {
      setServerStatus('checking');
      const healthData = await healthService.checkHealth();
      setServerStatus('online');
      setServerInfo(healthData);
    } catch (error) {
      setServerStatus('offline');
      Alert.alert(
        'Erreur de connexion',
        'Impossible de contacter le serveur FastAPI. Vérifiez qu\'il est lancé.'
      );
    }
  };

  const getStatusColor = () => {
    if (networkMode === 'offline') return '#e67e22'; // Orange pour mode hors ligne
    switch (serverStatus) {
      case 'online': return '#27ae60';
      case 'offline': return '#e74c3c';
      case 'checking': return '#f39c12';
      default: return '#95a5a6';
    }
  };

  return (
    <ScrollView style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.title}>SWMM Platform Mobile</Text>
        <Text style={styles.subtitle}>Outil terrain d'assainissement</Text>
      </View>

      {/* Switch Mode Réseau */}
      <View style={styles.networkSwitchCard}>
        <View style={styles.switchHeader}>
          <Text style={styles.switchTitle}>🔌 Mode Réseau Simulré</Text>
          <Switch
            value={networkMode === 'online'}
            onValueChange={toggleNetworkMode}
            trackColor={{ false: '#7f8c8d', true: '#2ecc71' }}
          />
        </View>
        <Text style={styles.switchDesc}>
          {networkMode === 'online'
            ? 'Mode connecté : Les données s\'envoient directement au serveur.'
            : 'Mode hors ligne : Toutes vos observations et incidents sont stockés localement.'}
        </Text>
      </View>

      {/* État du serveur */}
      <View style={styles.statusCard}>
        <Text style={styles.statusLabel}>État du système :</Text>
        <View style={[styles.statusIndicator, { backgroundColor: getStatusColor() }]} />
        <Text style={[styles.statusText, { color: getStatusColor() }]}>
          {networkMode === 'offline' ? 'Application Hors ligne' :
           serverStatus === 'online' ? 'Serveur en ligne ✓' :
           serverStatus === 'offline' ? 'Serveur injoignable ✗' : 'Vérification...'}
        </Text>
        
        {serverInfo && networkMode === 'online' && (
          <View style={styles.serverInfo}>
            <Text style={styles.infoText}>Base DB : {serverInfo.database}</Text>
            <Text style={styles.infoText}>Cache Graphe : {serverInfo.graph_cache}</Text>
          </View>
        )}
        
        {networkMode === 'online' && (
          <TouchableOpacity style={styles.refreshButton} onPress={checkServerStatus}>
            <Text style={styles.refreshButtonText}>Actualiser</Text>
          </TouchableOpacity>
        )}
      </View>

      {/* Menu Principal */}
      <View style={styles.menu}>
        <TouchableOpacity
          style={styles.menuItem}
          onPress={() => navigation.navigate('Map')}
        >
          <View style={styles.menuItemHeader}>
            <Text style={styles.menuItemTitle}>🗺️ Carte & Collecte</Text>
            {pendingCount > 0 && (
              <View style={styles.badge}>
                <Text style={styles.badgeText}>{pendingCount}</Text>
              </View>
            )}
          </View>
          <Text style={styles.menuItemDescription}>
            Fiches terrain, photos, GPS et signalement d'incidents
          </Text>
        </TouchableOpacity>

        <TouchableOpacity
          style={styles.menuItem}
          onPress={() => navigation.navigate('Sync')}
        >
          <View style={styles.menuItemHeader}>
            <Text style={styles.menuItemTitle}>🔄 Synchronisation</Text>
            {pendingCount > 0 && (
              <View style={styles.badgeOrange}>
                <Text style={styles.badgeText}>{pendingCount}</Text>
              </View>
            )}
          </View>
          <Text style={styles.menuItemDescription}>
            Synchroniser les données locales vers le bureau d'études
          </Text>
        </TouchableOpacity>
      </View>

      <View style={styles.footer}>
        <Text style={styles.footerText}>Version mobile 1.0.0 -- Dr A. Kellouche</Text>
      </View>
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: '#f5f6fa' },
  header: { backgroundColor: '#16213e', padding: 24, alignItems: 'center', borderBottomLeftRadius: 16, borderBottomRightRadius: 16 },
  title: { fontSize: 22, fontWeight: 'bold', color: '#fff' },
  subtitle: { fontSize: 13, color: '#bdc3c7', marginTop: 4 },
  networkSwitchCard: { backgroundColor: '#fff', margin: 16, padding: 16, borderRadius: 12, elevation: 2 },
  switchHeader: { flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center', marginBottom: 8 },
  switchTitle: { fontSize: 15, fontWeight: 'bold', color: '#2c3e50' },
  switchDesc: { fontSize: 12, color: '#7f8c8d', lineHeight: 18 },
  statusCard: { backgroundColor: '#fff', marginHorizontal: 16, marginBottom: 16, padding: 16, borderRadius: 12, elevation: 2, alignItems: 'center' },
  statusLabel: { fontSize: 14, fontWeight: 'bold', color: '#7f8c8d', marginBottom: 8 },
  statusIndicator: { width: 14, height: 14, borderRadius: 7, marginBottom: 6 },
  statusText: { fontSize: 15, fontWeight: 'bold', marginBottom: 8 },
  serverInfo: { width: '100%', borderTopWidth: 1, borderTopColor: '#f1f2f6', paddingTop: 8, marginTop: 8 },
  infoText: { fontSize: 12, color: '#57606f', marginBottom: 2 },
  refreshButton: { backgroundColor: '#16213e', paddingHorizontal: 16, paddingVertical: 8, borderRadius: 6, marginTop: 8 },
  refreshButtonText: { color: '#fff', fontWeight: 'bold', fontSize: 12 },
  menu: { padding: 16 },
  menuItem: { backgroundColor: '#fff', padding: 16, borderRadius: 12, marginBottom: 12, elevation: 2 },
  menuItemHeader: { flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center', marginBottom: 4 },
  menuItemTitle: { fontSize: 16, fontWeight: 'bold', color: '#16213e' },
  menuItemDescription: { fontSize: 12, color: '#7f8c8d', lineHeight: 16 },
  badge: { backgroundColor: '#2ecc71', paddingHorizontal: 8, paddingVertical: 2, borderRadius: 10 },
  badgeOrange: { backgroundColor: '#e67e22', paddingHorizontal: 8, paddingVertical: 2, borderRadius: 10 },
  badgeText: { color: '#fff', fontSize: 11, fontWeight: 'bold' },
  footer: { alignItems: 'center', marginTop: 24, marginBottom: 30 },
  footerText: { fontSize: 11, color: '#a4b0be', fontStyle: 'italic' },
});

File: mobile/lib/screens/home_screen.dart
import 'package:flutter/material.dart';
import 'package:swmm_mobile/services/api_service.dart';
import 'package:swmm_mobile/services/storage_service.dart';

class HomeScreen extends StatefulWidget {
  const HomeScreen({super.key});

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  String networkMode = 'online';
  String serverStatus = 'checking';
  Map<String, dynamic>? serverInfo;
  int pendingCount = 0;

  @override
  void initState() {
    super.initState();
    loadSettings();
  }

  Future<void> loadSettings() async {
    try {
      final mode = await storageService.getNetworkMode();
      setState(() => networkMode = mode);
      
      final changes = await storageService.getPendingChanges();
      setState(() => pendingCount = changes.length);

      if (mode == 'online') {
        checkServerStatus();
      } else {
        setState(() => serverStatus = 'offline');
      }
    } catch (e) {
      debugPrint(e.toString());
    }
  }

  Future<void> toggleNetworkMode(bool value) async {
    final newMode = value ? 'online' : 'offline';
    setState(() => networkMode = newMode);
    await storageService.setNetworkMode(newMode);
    
    if (newMode == 'online') {
      checkServerStatus();
    } else {
      setState(() {
        serverStatus = 'offline';
        serverInfo = null;
      });
    }
  }

  Future<void> checkServerStatus() async {
    try {
      setState(() => serverStatus = 'checking');
      final healthData = await HealthService().checkHealth();
      setState(() {
        serverStatus = 'online';
        serverInfo = healthData;
      });
    } catch (error) {
      setState(() => serverStatus = 'offline');
      if (mounted) {
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(content: Text('Impossible de contacter le serveur FastAPI')),
        );
      }
    }
  }

  Color getStatusColor() {
    if (networkMode == 'offline') return const Color(0xFFe67e22);
    switch (serverStatus) {
      case 'online': return const Color(0xFF27ae60);
      case 'offline': return const Color(0xFFe74c3c);
      case 'checking': return const Color(0xFFf39c12);
      default: return const Color(0xFF95a5a6);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SingleChildScrollView(
        child: Column(
          children: [
            Container(
              color: const Color(0xFF16213e),
              padding: const EdgeInsets.all(24),
              child: const Column(
                children: [
                  Text(
                    'SWMM Platform Mobile',
                    style: TextStyle(
                      fontSize: 22,
                      fontWeight: FontWeight.bold,
                      color: Colors.white,
                    ),
                  ),
                  SizedBox(height: 4),
                  Text(
                    'Outil terrain d\'assainissement',
                    style: TextStyle(
                      fontSize: 13,
                      color: Color(0xFFbdc3c7),
                    ),
                  ),
                ],
              ),
            ),
            Card(
              margin: const EdgeInsets.all(16),
              child: Padding(
                padding: const EdgeInsets.all(16),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Row(
                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
                      children: [
                        const Text(
                          '🔌 Mode Réseau Simulé',
                          style: TextStyle(
                            fontSize: 15,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                        Switch(
                          value: networkMode == 'online',
                          onChanged: toggleNetworkMode,
                          activeColor: const Color(0xFF2ecc71),
                        ),
                      ],
                    ),
                    const SizedBox(height: 8),
                    Text(
                      networkMode == 'online'
                          ? 'Mode connecté : Les données s\'envoient directement au serveur.'
                          : 'Mode hors ligne : Toutes vos observations et incidents sont stockés localement.',
                      style: const TextStyle(
                        fontSize: 12,
                        color: Color(0xFF7f8c8d),
                      ),
                    ),
                  ],
                ),
              ),
            ),
            Card(
              margin: const EdgeInsets.symmetric(horizontal: 16),
              child: Padding(
                padding: const EdgeInsets.all(16),
                child: Column(
                  children: [
                    const Text(
                      'État du système :',
                      style: TextStyle(
                        fontSize: 14,
                        fontWeight: FontWeight.bold,
                        color: Color(0xFF7f8c8d),
                      ),
                    ),
                    const SizedBox(height: 8),
                    Container(
                      width: 14,
                      height: 14,
                      decoration: BoxDecoration(
                        color: getStatusColor(),
                        shape: BoxShape.circle,
                      ),
                    ),
                    const SizedBox(height: 6),
                    Text(
                      networkMode == 'offline'
                          ? 'Application Hors ligne'
                          : serverStatus == 'online'
                              ? 'Serveur en ligne ✓'
                              : serverStatus == 'offline'
                                  ? 'Serveur injoignable ✗'
                                  : 'Vérification...',
                      style: TextStyle(
                        fontSize: 15,
                        fontWeight: FontWeight.bold,
                        color: getStatusColor(),
                      ),
                    ),
                    if (serverInfo != null && networkMode == 'online') ...[
                      const SizedBox(height: 8),
                      const Divider(),
                      Text(
                        'Base DB : ${serverInfo!['database']}',
                        style: const TextStyle(fontSize: 12),
                      ),
                      Text(
                        'Cache Graphe : ${serverInfo!['graph_cache']}',
                        style: const TextStyle(fontSize: 12),
                      ),
                    ],
                    if (networkMode == 'online') ...[
                      const SizedBox(height: 12),
                      ElevatedButton(
                        onPressed: checkServerStatus,
                        style: ElevatedButton.styleFrom(
                          backgroundColor: const Color(0xFF16213e),
                        ),
                        child: const Text(
                          'Actualiser',
                          style: TextStyle(color: Colors.white),
                        ),
                      ),
                    ],
                  ],
                ),
              ),
            ),
            Padding(
              padding: const EdgeInsets.all(16),
              child: Column(
                children: [
                  Card(
                    child: InkWell(
                      onTap: () => Navigator.pushNamed(context, '/map'),
                      child: Padding(
                        padding: const EdgeInsets.all(16),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                          children: [
                            const Column(
                              crossAxisAlignment: CrossAxisAlignment.start,
                              children: [
                                Text(
                                  '🗺️ Carte & Collecte',
                                  style: TextStyle(
                                    fontSize: 16,
                                    fontWeight: FontWeight.bold,
                                  ),
                                ),
                                SizedBox(height: 4),
                                Text(
                                  'Fiches terrain, photos, GPS et signalement d\'incidents',
                                  style: TextStyle(
                                    fontSize: 12,
                                    color: Color(0xFF7f8c8d),
                                  ),
                                ),
                              ],
                            ),
                            if (pendingCount > 0)
                              Container(
                                padding: const EdgeInsets.symmetric(
                                    horizontal: 8, vertical: 2),
                                decoration: BoxDecoration(
                                  color: const Color(0xFF2ecc71),
                                  borderRadius: BorderRadius.circular(10),
                                ),
                                child: Text(
                                  '$pendingCount',
                                  style: const TextStyle(
                                    color: Colors.white,
                                    fontSize: 11,
                                    fontWeight: FontWeight.bold,
                                  ),
                                ),
                              ),
                          ],
                        ),
                      ),
                    ),
                  ),
                  const SizedBox(height: 12),
                  Card(
                    child: InkWell(
                      onTap: () => Navigator.pushNamed(context, '/sync'),
                      child: Padding(
                        padding: const EdgeInsets.all(16),
                        child: Row(
                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                          children: [
                            const Column(
                              crossAxisAlignment: CrossAxisAlignment.start,
                              children: [
                                Text(
                                  '🔄 Synchronisation',
                                  style: TextStyle(
                                    fontSize: 16,
                                    fontWeight: FontWeight.bold,
                                  ),
                                ),
                                SizedBox(height: 4),
                                Text(
                                  'Synchroniser les données locales vers le bureau d\'études',
                                  style: TextStyle(
                                    fontSize: 12,
                                    color: Color(0xFF7f8c8d),
                                  ),
                                ),
                              ],
                            ),
                            if (pendingCount > 0)
                              Container(
                                padding: const EdgeInsets.symmetric(
                                    horizontal: 8, vertical: 2),
                                decoration: BoxDecoration(
                                  color: const Color(0xFFe67e22),
                                  borderRadius: BorderRadius.circular(10),
                                ),
                                child: Text(
                                  '$pendingCount',
                                  style: const TextStyle(
                                    color: Colors.white,
                                    fontSize: 11,
                                    fontWeight: FontWeight.bold,
                                  ),
                                ),
                              ),
                          ],
                        ),
                      ),
                    ),
                  ),
                ],
              ),
            ),
            const SizedBox(height: 24),
            Text(
              'Version mobile 1.0.0 -- Dr A. Kellouche',
              style: const TextStyle(
                fontSize: 11,
                color: Color(0xFFa4b0be),
                fontStyle: FontStyle.italic,
              ),
            ),
          ],
        ),
      ),
    );
  }
}